code/src/sixsq/nuvla/server/resources/nuvlabox_playbook.clj

```diff
@@ -0,0 +1,169 @@
+(ns sixsq.nuvla.server.resources.nuvlabox-playbook
+  "
+The nuvlabox-playbook resource holds scripts that, when configured, 
+can be executed by the NuvlaBox device, externally and independently of the 
+NuvlaBox Engine software
+"
+  (:require
+    [clojure.tools.logging :as log]
+    [sixsq.nuvla.auth.acl-resource :as a]
+    [sixsq.nuvla.auth.utils :as auth]
+    [sixsq.nuvla.db.impl :as db]
+    [sixsq.nuvla.server.resources.common.crud :as crud]
+    [sixsq.nuvla.server.resources.common.std-crud :as std-crud]
+    [sixsq.nuvla.server.resources.common.utils :as u]
+    [sixsq.nuvla.server.resources.job :as job]
+    [sixsq.nuvla.server.resources.nuvlabox.utils :as utils]
+    [sixsq.nuvla.server.resources.resource-metadata :as md]
+    [sixsq.nuvla.server.resources.spec.nuvlabox-playbook :as nb-playbook]
+    [sixsq.nuvla.server.util.log :as logu]
+    [sixsq.nuvla.server.util.metadata :as gen-md]
+    [sixsq.nuvla.server.util.response :as r]))
+
+
+(def ^:const resource-type (u/ns->type *ns*))
+
+
+(def ^:const collection-type (u/ns->collection-type *ns*))
+
+
+(def collection-acl {:add   ["group/nuvla-user"]
+                     :query ["group/nuvla-user"]})
+
+
+;;
+;; validation
+;;
+
+(def validate-fn (u/create-spec-validation-fn ::nb-playbook/schema))
+
+
+(defmethod crud/validate resource-type
+  [resource]
+  (validate-fn resource))
+
+;;
+;; acl
+;;
+
+(defmethod crud/add-acl resource-type
+  [resource request]
+  (if-let [nuvlabox-id (:parent resource)]
+    (let [{nuvlabox-acl :acl} (crud/retrieve-by-id-as-admin nuvlabox-id)
+          view-acl (:view-acl nuvlabox-acl)
+          edit-acl (:edit-acl nuvlabox-acl)]
+      (assoc resource
+             :acl (cond-> (assoc (or (:acl resource) (a/default-acl (auth/current-authentication request))) :manage [nuvlabox-id])
+                    (not-empty view-acl) (assoc :view-acl (into [] (distinct (merge view-acl nuvlabox-id))))
+                    (not-empty edit-acl) (assoc :edit-acl edit-acl))))
+    (a/add-acl resource request)))
+
+;;
+;; CRUD operations
+;;
+
+(def add-impl (std-crud/add-fn resource-type collection-acl resource-type))
+
+
+(defmethod crud/add resource-type
+  [{{:keys [parent] :as body} :body :as request}]
+  (some-> parent
+    (db/retrieve request)
+    (a/throw-cannot-edit request))
+  (-> request
+    (update-in [:body] dissoc :output)
+    (add-impl)))
+
+
+(def edit-impl (std-crud/edit-fn resource-type))
+
+
+(defmethod crud/edit resource-type
+  [{:keys [body] :as request}]
+  (if-let [updated-output (:output body)]
+    (edit-impl (assoc request :body (assoc body :output (subs updated-output 0 (min (count updated-output) 1000)))))
```

Comment:
How about this implementation. I suspect that the limit-string-size would be useful in other places and we can add it to utils.
```
(defn limit-string-size
  [limit s]
  (cond-> s
          (> (count s) limit) (subs 0 limit)))
  
(defmethod crud/edit resource-type
  [{{:keys [output]} :body :as request}]
  (-> request
        (cond-> output (assoc-in [:body :output] (limit-string-size 1000 output)))
        edit-impl))
```

Reply:
done